### PR TITLE
Add statistics endpoint and service

### PR DESCRIPTION
**Pull request by deen13**
There are two code smells left which i was unable to resolve. :confounded:

```
Ruleset: naming - 5min debt
        MatchingDeclarationName - [Configuration.kt] at /Users/dennisdierkes/Github/kickway/src/main/kotlin/de/smartsquare/kickchain/kickway/Configuration.kt:1:1
Ruleset: performance
Ruleset: potential-bugs
Ruleset: style - 10min debt
        UtilityClassWithPublicConstructor - [KickchainException] at /Users/dennisdierkes/Github/kickway/src/main/kotlin/de/smartsquare/kickchain/kickway/analyzing/Exceptions.kt:3:1
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -78,6 +78,8 @@ dependencies {
     implementation "org.springframework.boot:spring-boot-starter-hateoas"
 
     implementation "com.fasterxml.jackson.module:jackson-module-kotlin:$jacksonVersion"
+    implementation "com.fasterxml.jackson.core:jackson-databind:$jacksonVersion"
+    implementation "com.fasterxml.jackson.core:jackson-annotations:$jacksonVersion"
```
Comment (rubengees) [inline build.gradle]: Aren't these included in Spring?

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,5 +1,6 @@
+#Tue Aug 21 08:17:30 CEST 2018
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.9-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.9-all.zip
```
Comment (rubengees) [inline gradle/wrapper/gradle-wrapper.properties:6]: IntelliJ suggestes this, but I disagree with this practice. The sources included in the `all`-jar are not worth the larger size in my opinion.

File: src/main/kotlin/de/smartsquare/kickchain/kickway/KickwayConfigurationProperties.kt [hunk 3]
```diff
@@ -23,5 +23,12 @@ class KickwayConfigurationProperties {
         @get:URL
         @get:NotBlank
         var url by Delegates.notNull<String>()
+
+        @get:NotBlank
+        var name by Delegates.notNull<String>()
+
+        @get:NotBlank
+        var password by Delegates.notNull<String>()
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/KickwayConfigurationProperties.kt]: Maybe put these into an object named `Authorization` to have more clear properties like: `kickway.kickchain.authorization.password`?

File: src/main/kotlin/de/smartsquare/kickchain/kickway/analyzing/BlockchainAnalyzer.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            } else {
+                listOf(
+                    it.team1.players.first() to it.score.goals1,
+                    it.team2.players.first() to it.score.goals2,
+                    it.team2.players[1] to it.score.goals2
+                )
+            }
+        }.topTen()
+
+
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/analyzing/BlockchainAnalyzer.kt]: Nit: Duplicate new line.

File: src/main/kotlin/de/smartsquare/kickchain/kickway/analyzing/BlockchainAnalyzer.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                )
+            }
+            .filter { (playerName) -> playerName == name }
+
+        if (games.isEmpty()) {
+            throw UnknownPlayerException(name)
+        }
+
+        val averageGoalsPerGame = games.sumBy { (_, goals) -> goals } / games.size
+        val winRate: Double = games.filter { (_, goals) -> goals == 10 }.size.toDouble().div(games.size)
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/analyzing/BlockchainAnalyzer.kt]: Maybe also total wins and total losses?

File: src/test/groovy/de/smartsquare/kickchain/kickway/analyzing/BlockchainAnalyzerSpecification.groovy [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                .where('deen').played(10).to(1).against('ruby')
+                .and('skonair').played(10).to(1).against('deen')
+                .and('alexn').played(10).to(7).against('ruby')
+                .and('deen').played(3).to(10).against('danielr')
+                .and('saschar').played(5).to(10).against('alexr')
+                .and('marco').played(10).to(8).against('natascha')
+                .and('drs').played(10).to(2).against('jensk')
+                .and('deen').played(10).to(1).against('ruby')
+                .and('deen').played(10).to(1).against('ruby')
+                .finalizedBy('deen').played(10).to(1).against('ruby')
```
Comment (rubengees) [inline src/test/groovy/de/smartsquare/kickchain/kickway/analyzing/BlockchainAnalyzerSpecification.groovy:24]: :100:

File: src/test/groovy/de/smartsquare/kickchain/kickway/analyzing/BlockchainBuilder.groovy [hunk 7]
```diff
@@ -0,0 +1,59 @@
+package de.smartsquare.kickchain.kickway.analyzing
+
+class BlockchainBuilder {
```
Comment (rubengees) [inline src/test/groovy/de/smartsquare/kickchain/kickway/analyzing/BlockchainBuilder.groovy:3]: This is great :clap:

File: src/test/groovy/de/smartsquare/kickchain/kickway/analyzing/KickchainRepositorySpecification.groovy [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                .withPath("/chain"), exactly(1))
+                .respond(response()
+                .withStatusCode(500))
+
+        when:
+        repository.fetch()
+
+        then:
+        def error = thrown(KickchainException)
+        error.message == 'org.springframework.web.client.HttpServerErrorException: 500 Internal Server Error'
```
Comment (rubengees) [inline src/test/groovy/de/smartsquare/kickchain/kickway/analyzing/KickchainRepositorySpecification.groovy]: We should use a better message here. Callers should not see internals of the technologies we are using.

File: src/main/kotlin/de/smartsquare/kickchain/kickway/analyzing/BlockchainRepository.kt [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package de.smartsquare.kickchain.kickway.analyzing
+
+interface BlockchainRepository {
+
+    fun fetch(): Blockchain
+
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/analyzing/BlockchainRepository.kt]: Nit: Unnecessary newline.

File: gradle/wrapper/gradle-wrapper.properties [hunk 10]
```diff
@@ -1,5 +1,6 @@
+#Tue Aug 21 08:17:30 CEST 2018
```
Comment (rubengees) [inline gradle/wrapper/gradle-wrapper.properties:1]: Where does this come from? Did you generate with `./gradlew wrapper`?

**Review by rubengees — CHANGES_REQUESTED** [review]
Nice! Just a few things.

File: src/main/kotlin/de/smartsquare/kickchain/kickway/Configuration.kt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package de.smartsquare.kickchain.kickway
+
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.boot.web.client.RestTemplateBuilder
+import org.springframework.web.client.RestTemplate
+
+@Configuration
+class HttpBasicAuthConfiguration(private val properties: KickwayConfigurationProperties) {
```
Comment (rubengees) [inline src/main/kotlin/de/smartsquare/kickchain/kickway/Configuration.kt]: The classname is different as the filename.
This is the first issue detekt reports.

**Comment by rubengees** [comment]
The `UtilityClassWithPublicConstructor` violation seems to be a false positive. You can ignore it with `@Suppress("UtilityClassWithPublicConstructor")`.

File: build.gradle (hunk 1 above)
Comment (deen13) [reply build.gradle]: You're right 👍

File: src/main/kotlin/de/smartsquare/kickchain/kickway/Configuration.kt (hunk 11 above)
Comment (deen13) [reply src/main/kotlin/de/smartsquare/kickchain/kickway/Configuration.kt]: That was a conscious decision. I would like to wrap all configuration class in the one Configuration.kt. Should i just suppress this warning?

Comment (rubengees) [reply src/main/kotlin/de/smartsquare/kickchain/kickway/Configuration.kt]: Resolved offline.

**Pull request merged**